### ISSUE TITLE
Storage nodes may not use reputation contract that much

### ISSUE THREAD

**Issue by alexvanin** [nspcc-dev/neofs-node#1037]
During reputation contract [debugging](https://github.com/nspcc-dev/neofs-contract/pull/201) I dumped testnet contract storage. It contains _35_ reputation records to this date. That's not much for 2221 epochs. All values from the contract storage presented there (note that neofs-contract v0.13.0 release provides new contract storage scheme for values, those hex-dumped values are basically N3 VM encoded `[][]byte`).

```
key : value
020296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d5 : 400128c50a040802100912550a230a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691122e0a230a210296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d511736891ed7c3fa53f1a660a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a496911241040b0308f19ec68c5bbaa40b93125051c44f795c8ceb645d36648b01cafd68a58c7f4f468e22a240e60092ec31b43c566713caad4ea6a8dc84834c1a0896939a3f
0203586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691 : 400128c50a040802100912550a230a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f122e0a230a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a4969111736891ed7c3fa53f1a660a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f1241047581d11fdf7382bc9e1cb96da980286a3b869bc55a95eb239a318f7f1c3cc3df158a0aa58f4a4015cd7f80fd185fc26275ac6ecd385a6869ba052c2115658024
16040319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f : 400128c50a040802100912550a230a210220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc2122e0a230a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f111ae25817b7d1a73f1a660a210220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc212410411de56e4b0d6cef44267c990aa06e0b10e08b591fff51f9fcdcda8b225b98b2baf092e5add676e22b4d3f380d421f91fbbda843fb2815f2b44f6b720b90c13f9
24030319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f : 400128c50a040802100912550a230a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691122e0a230a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f117a5bd6ea98e0b53f1a660a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691124104d84a03868eba6499be227ec5feb9843c182d9aa254782863feef171057209825f6c1f736e2a8b99a9dd5f51511c8d201aacd39c516403990c4f4bbc86da10f5c
3f030319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f : 400128c50a040802100912550a230a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691122e0a230a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f114447939ab6d69e3f1a660a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a4969112410483da7f741d18ed59eeecfb0fc508d5a0f588a3dba88ab18ba7a6fb17a46481ab4102521e199edf0fb7d907c20847e99b85d6e66b88d47163d5612380a4b92c42
40030220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc2 : 400128c50a040802100912550a230a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c08090122e0a230a210220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc2114ea60a4625759c3f1a660a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c080901241046f2beabc7d6ff1f1f62528e12b80472ea217286365d031faaf50123f3dea54b2975ea4af8b486144646dea4d594c3a66407c5af33fcfa104233556637bf51b76
40030319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f : 400128c50a040802100912550a230a210220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc2122e0a230a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f114447939ab6d69e3f1a660a210220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc212410491fb8e05842b331a3b5c4e5b7296322aea03329a89e5e7b982ce3e4427d818693fab6532aaf0f475be7cdc4d78a75bd41457032604bee8a17d69b03e58937449
41030220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc2 : 400128c50a040802100912550a230a210296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d5122e0a230a210220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc211f45ffa710eb1943f1a660a210296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d5124104ca260103aa9efa2a73821ec641ac8200784761a34a7ccbc5565c9d190ca95518ff6da3d8f803018870cf1cbb8a2bd9bdf5cdb3cefebf74b67feb9736802c2cb1
410302dab9ce2e8ca427c3932fbb5cbb0748298357639331a5fe9f0c644b1dc953f9bc : 400128c50a040802100912550a230a210220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc2122e0a230a2102dab9ce2e8ca427c3932fbb5cbb0748298357639331a5fe9f0c644b1dc953f9bc11f45ffa710eb1943f1a660a210220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc212410412dd643536b7f27a8327550475eabe0bf2d26229d545006eec2adf18d14338de95d43fd768b003917697c18355f730cf7ebcbc526de6b181781554cca23fa8c1
42030220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc2 : 400128c50a040802100912550a230a210296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d5122e0a230a210220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc211f45ffa710eb1943f1a660a210296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d5124104bf4b91d0314416a9caf8c4b1501bc7bed6b8551e885cd6b70458aab970cd13fc1f9b1b49c1196461292d647bf61da5d36f84fc489227c22634e0fb80d98b5cc5
420302dab9ce2e8ca427c3932fbb5cbb0748298357639331a5fe9f0c644b1dc953f9bc : 400128c50a040802100912550a230a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691122e0a230a2102dab9ce2e8ca427c3932fbb5cbb0748298357639331a5fe9f0c644b1dc953f9bc11f45ffa710eb1943f1a660a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691124104c9f48df2dec18ec75b32b68daf901c97af66c056169cfde68d8f4c04255d27e37edfb19919346e8f3d5a4d70f84af619127acc3e16a7c772032f73b8ebc51beb
6a0202bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a : 400128c50a040802100912550a230a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c08090122e0a230a2102bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a11738a8ee4f21f9a3f1a660a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c0809012410418d762b0c47e0766cef6731e6e02642b168a999a3de49c42a825d9e0bb23a632a5714e04263636a58c48ea4878386b24badddefac074ab0e6f188d38237470a4
6a020319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f : 400128c50a040802100912550a230a210296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d5122e0a230a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f11d6cc317163899d3f1a660a210296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d51241048af963d66cf5cf82e177c059d545a9a3b913cf4dc78b6366f2eb5f9ff9e0de3f1e3bfe6e562570ce7821a1195390d5a136143f0b6088e1a4ef4d1a39ad7de4ae
6e050259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c08090 : 400128c50a040802100a12550a230a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691122e0a230a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c0809011b9caa8029fc2ab3f1a660a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691124104c49b7bcbab9a936bfb0294a7ed20daf44810d6bf9981d81db19793957c651f2a543d2ff5272a9c2e46eb41ddf80304f388f57a85a236ff3cca803c769f0ea129
6e050296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d5 : 400128c50a040802100a12550a230a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f122e0a230a210296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d511b29defa7c64bab3f1a660a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f1241043429789567d29b4ba55aefe498eb306f331d7acfe7750cc3374bd1c75739dc682767684ae49346a240af92b64a81dd87f7b7acc57f9cec6f6dccc5a0e2c89a0a
800202bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a : 400128c50a040802100912550a230a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691122e0a230a2102bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a114ea60a4625759c3f1a660a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a496911241044efb85852f5ee534efaba2c7f46806f7b7c71b0ba2fe0a2d9fdea7074642ce996ea18872ee107a0c14b98886f7af5f01f609c64a2b4e2663ac3fdf1babe4627c
80020319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f : 400128c50a040802100912550a230a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c08090122e0a230a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f1186309d7703bb9d3f1a660a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c0809012410496cb962c259ef04c5ec9528ec58f4e25e41abba9206f8a95d7c94c52710a2b9bef62060cfb6e8bfdd6280438b3ce7a805eb6291c66d5da3ec4c7045672ae96be
810202bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a : 400128c50a040802100912550a230a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f122e0a230a2102bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a114ea60a4625759c3f1a660a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f12410439e79fd5b4564252706351378472d11eb2dc42e9eae859cd445555f92c407239b400a75bbd49a55249518ab8abdddce16558fb6a48fc466813e9a59b08d0fd5a
81020319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f : 400128c50a040802100912550a230a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c08090122e0a230a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f111f6f216bff449a3f1a660a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c08090124104f5fe1e7a54c7c155dd0422cee69a1a71b9cbc68cbc89d7138ffe050a78de18e8cbaaef104ef7d63994f4abf41abff44c261de3fc11bb425bb5d74286b43bca41
950202bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a : 400128c50a040802100912550a230a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c08090122e0a230a2102bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a11738a8ee4f21f9a3f1a660a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c080901241043806e01f39f6267f2b508e0c261f003e33aa78466a480c2105ef7bb347602478cefa718a937f40f2024e26255d5cda151d4122e6cd91efd688bcf55090709547
950203586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691 : 400128c50a040802100912550a230a210296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d5122e0a230a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691110835828f6f9c9e3f1a660a210296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d5124104ebee3486910c7e57c9bdaf620d08561cc7306daafea7aabc9b4eac7be8119d5cd9936c39e9e810020ac9a700602ca1dcb6ad2167d3793f88a8e4dcd7547d08c8
97000319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f : 400128c50a040802100912550a230a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c08090122e0a230a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f11fca9f1d24d62b03f1a660a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c08090124104c8e1f6e7ad52ee2e18697eb23e1c8467db3b05ba23e8d50eacf8c2bab1c7dfd8ccd2dffa1e9a46e5956a9e2f844f417ea3c36bedbffe25f722954cfd3259a4e7
c50202bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a : 400128c50a040802100912550a230a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c08090122e0a230a2102bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a11738a8ee4f21f9a3f1a660a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c08090124104e60a33a1d805cb68bf53a64842f6b5d8d729b2eb019b422c70fb226fad2c9b1aeae50e5ab1820499d1c20e7db8a8cfba72fe7706f354e7951baab0f7cf3b1b52
c5020319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f : 400128c50a040802100912550a230a2102bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a122e0a230a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f1104cdcd699a939c3f1a660a2102bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a124104544aba9c63af1a8ec3f57d0a1a6a6eb8597098ec8fccbc3f0a0035ca016081b59e4a7d46868fb28aced834181d6dbd9d529900a2c8181c7b941827d394194828
c7020296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d5 : 400128c50a040802100912550a230a2102bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a122e0a230a210296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d5119228c89bcad99e3f1a660a2102bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a1241048b548fdd429ce8ad11b831f30285c98ad5bbbf2d3d233c7661703944e42260981f3b13d501f1a1f50e881d7ccca076a10fc2123682913fbe37c7b256235c223b
c7020319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f : 400128c50a040802100912550a230a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691122e0a230a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f1129b15108683e9a3f1a660a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691124104d27cded5d77586fdeab507c586c5718dc7c77d0389ca7f39311e584600727821a778b447ba12481c04e06e81f48fd6cdf6f72fdac1ac5506021e319768d32a27
c8020296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d5 : 400128c50a040802100912550a230a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c08090122e0a230a210296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d511e2adb267b73ca73f1a660a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c08090124104de383ca892c03eeeea4aeffb206376185829c2da08caa3c4843f89dcd622b431be857c7ae8b9230e450c5df8ba1764bb00f89ea5e5c621cafc97df287d577523
c80202bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a : 400128c50a040802100912550a230a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691122e0a230a2102bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a119115a1c95b13a73f1a660a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691124104c34d124ce573f3ca2d59515cac5057ad278f5da8ceea85b55cd2619e1727fd60b5cabd6895965e19929d405725a0fe10cc0b017860b6bed66e8e945ba2b629b7
c8020319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f : 400128c50a040802100912550a230a2102bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a122e0a230a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f11517d265d6557a83f1a660a2102bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a124104797843e033da30fc86bea6bd8f4683bac9496bb8ccd221ba9308f24a1caca5a8b08be588df458e47e2e83b6f2202871c1dd2503114fcdf059e499b830354dbe0
d5070319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f : 400128c50a040802100a12550a230a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c08090122e0a230a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f114ea60a4625759c3f1a660a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c0809012410431d4e859210f1119c8fbc70c02169ed746410e5cb152af462b37aaf16564217b4dd48d8638237f36a382d73246ca73621ad7a097c4c9d6c31134a7385796bf31
d6070319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f : 400128c50a040802100a12550a230a210296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d5122e0a230a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f113e8724c0fdaa9c3f1a660a210296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d51241049c9e03bb21fc4ae0d52941640128a847348a89038d20aff133aa0db85f3741a48f8179d8abd5d9dcb291cec1a3587be789e5bf7153602b413da66cd452b03fbb
d60703586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691 : 400128c50a040802100a12550a230a210220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc2122e0a230a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691119c73c0d7d1019f3f1a660a210220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc2124104ad1be738e9183beb4a5f782d584fe3f885571c547497d304b09130285ac1120cae418b00468fd6e4c84b3a5d272c79997bf606248170dcd3a77d063beeca5704
e5030319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f : 400128c50a040802100912550a230a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c08090122e0a230a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f11c620b07268919f3f1a660a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c0809012410487d91a3eef3055fffd2f8a3148cbb76947f7d399a27f0dd455fc0f5462b7052313b430b60b096e30f032e97353fc9f5a5edb22aaf9ead34ff0eeb4c7ca294641
ff030220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc2 : 400128c50a040802100a12550a230a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c08090122e0a230a210220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc2113d41290c6cae9c3f1a660a210259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c080901241048dd30484e21f358839ef91cb2d00a386a92b606eade604ac9455249ae5d20062445cbcf1535e73552322d3dbab99543cb4b395610191926fb22c389c1427e8b8
ff0303586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691 : 400128c50a040802100a12550a230a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f122e0a230a2103586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691119247f0e953729e3f1a660a210319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f124104855a394c4d3802fef38fccc7e42fab72f996174f4b81d9a3d92edd6b29e31e826b5ea6ad971765c23f78b031bbf29b807655b2d51d413a0b3db8a79d2dbb085a
```

**Comment by carpawell** [comment]
I dumped (via RPC) storage KV pairs (reputation + service information) for `mainnet`, `testnet`, `dev-env` (was up for ~3h with default 5 minutes epoch).

Results:

| Chain | Contract K/V Number |
| -------- | ---------------------------- |
| MainNet | 574 |
| TestNet | 76 |
| DevEnv | 233 |

In `dev-env` storage IR has not produced any errors related to reputation values but has logged debug info about sending global trusts to the contract consistently once per epoch. I tried to stop IR/SN a few times to check if it breaks something but reputation values continued to be written in the contract's storage.

TestNet has neither debug logs (are turned off) nor errors about sending, but its storage volume has increased since issue creation, so at least it is not constant and is increasing. However, values in the contract are strange and do not have any obvious logic:

```
epoch 2, counter 1, peer: 0296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d5
epoch 2, counter 1, peer: 03586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691
epoch 1046, counter 1, peer: 0319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f
epoch 2841, counter 1, peer: 02dab9ce2e8ca427c3932fbb5cbb0748298357639331a5fe9f0c644b1dc953f9bc
epoch 2841, counter 1, peer: 03641c726c98cd6ff3c1b919f425c9d67b5702bc7c7ffe6bc8d8d86b8fdb3c80bc
epoch 804, counter 1, peer: 0319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f
epoch 831, counter 1, peer: 0319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f
epoch 832, counter 1, peer: 0220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc2
epoch 832, counter 1, peer: 0319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f
epoch 833, counter 1, peer: 0220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc2
epoch 833, counter 1, peer: 02dab9ce2e8ca427c3932fbb5cbb0748298357639331a5fe9f0c644b1dc953f9bc
epoch 834, counter 1, peer: 0220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc2
epoch 834, counter 1, peer: 02dab9ce2e8ca427c3932fbb5cbb0748298357639331a5fe9f0c644b1dc953f9bc
epoch 618, counter 1, peer: 02bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a
epoch 618, counter 1, peer: 0319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f
epoch 1390, counter 1, peer: 0259ecf08766715ed43a05573f265baa5ba7664c7b9f59f22f8efd018906c08090
epoch 1390, counter 1, peer: 0296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d5
epoch 640, counter 1, peer: 02bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a
epoch 640, counter 1, peer: 0319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f
epoch 641, counter 1, peer: 02bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a
epoch 641, counter 1, peer: 0319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f
epoch 661, counter 1, peer: 02bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a
epoch 661, counter 1, peer: 03586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691
epoch 151, counter 1, peer: 0319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f
epoch 709, counter 1, peer: 02bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a
epoch 709, counter 1, peer: 0319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f
epoch 711, counter 1, peer: 0296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d5
epoch 711, counter 1, peer: 0319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f
epoch 712, counter 1, peer: 0296416a8da7e2d144688d3f1f7af9321b88d12b0497c9b630d92a0f483610f8d5
epoch 712, counter 1, peer: 02bab7cb47d7c493d84e4400443d99ccfb451db8503615bb2b2c752f649632362a
epoch 712, counter 1, peer: 0319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f
epoch 2005, counter 1, peer: 0319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f
epoch 2006, counter 1, peer: 0319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f
epoch 2006, counter 1, peer: 03586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691
epoch 997, counter 1, peer: 0319934feaba77dfd00dfab978ccffa032b94795eda0bee11436b05d907b8c0f7f
epoch 1023, counter 1, peer: 0220f54c504a89f06314d9a94287ef9bd77763d18591c53c55c5867c5772d75cc2
epoch 1023, counter 1, peer: 03586cfc335cde71d96d99aff5b1cf7f4d039a55014a5d202da39d5d17a5a49691
total values count: 37
```
It seems that reputation values started to be a problem at the beginning of the network (of course if it is considered that no values have been deleted by a mistake in a migration). Also, there are 8 different trusted peers in the contract memory, not sure if it is ok.

Test that prints reputation values: https://gist.github.com/carpawell/184514fc1ef6ebf2ed5f8423623c6933 (note that `neo-go` library has to [contain](https://github.com/nspcc-dev/neo-go/pull/2328) `FindStates` method)

**Comment by carpawell** [comment]
TL;DR: The reputation process uses two *independent* events: `NewBlock` notification, `NewEpoch` event. It seems that we have some kind of race in the reputation process.

### Detailed.

The reputation component works with its own "time" based on block count in the morph chain. It is used for [defining](https://github.com/nspcc-dev/neofs-node/blob/v0.27.5/cmd/neofs-node/timers.go#L61) when it is time to run the next iteration of the Eigen Trust algorithm, e.g.: epoch duration is 20 and number of iterations is 4 => move to the next iteration every 20/4=5 blocks and the last iteration should write reputation values to the contract.

However, we have to handle `NewEpoch` event from the morph to be flexible and be able to adopt timers for an epoch duration or/and a number of iterations changes. Now it works as a [sync](https://github.com/nspcc-dev/neofs-node/blob/v0.27.5/cmd/neofs-node/reputation.go#L229) and resets reputation timer. So basically if the `NewEpoch` event comes in the next block (after the real NewEpoch block) we reset our timers and start working with one block offset. The next `NewEpoch` event would be received before the last Eigen Trust iteration. This leads to the early timers reset and does not allow any reputation values to be written to the contract (the last iteration of the current epoch is never processed).

Using worker pools for both processes also complicates the situation: even if the order of the call was right, the results may differ depending on the execution of the worker (I saw such examples).

I tried to use more iterations per epoch (the number is the same `n` but timers would try to tick `n+1` times), it helped a little, but brought another problem: if "extra" iteration was accepted "at the wrong time", it breaks controller since it starts living in their own world. The process that I saw:
1. now is the `n-th` epoch
2. \*new epoch\*
3. submit resetting timer to the worker pool
4. submit ticking extra iteration (it now thinks that it is `n+1` epoch via `RPC` call) to the worker pool
5. actual early iteration handling (`n+1-th` epoch, `0` iteration), this is an error, no information is available in the node
6. actual timer resetting
7. \*all system thinks that the next tick will be (`n+1-th`, `0`) but the current node will do (`n+1-th`, `1`) and will not have information about iterations that were not handled by other nodes yet\*

### Current approach

So to solve this *at least* two addition should be implemented:
1. some sort of sync between processes to make all iterations be called synchronously with respect to this `NewEpoch` event handling (looks easy, but not that easy in fact: it is hard to understand what is "the last iteration" that we should wait for in a general sense, what if it was unplanned `NewEpoch` event in the middle of the current epoch, how not to get deadlock, etc)
2. add "no-op" iteration support: we have to keep in mind that some problems may appear and an unexpected `Tick()` call may be done and break some logic in the execution. That is easy but sounds weird IMO.

### Alternatives

@cthulhu-rider suggested two alternatives:
1. Do not make iterations in the middle of an epoch (drop timers for the reputation): all the iterations could be done on `NewEpoch` event. So if a number of the iterations in the algo is `i` and the current epoch is `n`, then every epoch nodes do `i` iterations (one of them is the last iteration for the `n-i-1` epoch and one of them is the initial iteration of the current epoch, others are the intermediate). But of course, reputation results would be written to the contract with `i` epoch delay.
2. Make `IR` the only source of truth that says when to start an iteration. This could be done via morph calls and notifications. But of course, this requires additional contract methods, RPC calls, notification handlers, and logic inside the IR.

@alexvanin, @realloc, @cthulhu-rider, @fyrchik, thoughts?

**Comment by realloc** [comment]
After talk with @carpawell , I suggest to define the minimal number of iterations to consider the result statistically acceptable and when a new Epoch comes, analyze the already gathered amount of reputation data. If it's not enough (Epoch was too short), drop the result, otherwise finish the calculation for the past Epoch and write the result. When a new Epoch starts, the statistics are accumulated separately, not to mix the operation periods, in an asynchronous way. Maybe, we need to separate the cycle time in blocks from the Epoch duration and look at block creation time with ability to adjust the cycle length to meet the minimal requirements, if possible.

**Comment by alexvanin** [comment]
Once again we discussed it with @carpawell and @cthulhu-rider. We come up mostly with the same approach from @realloc  

> Maybe, we need to separate the cycle time in blocks from the Epoch duration and look at block creation time with ability to adjust the cycle length to meet the minimal requirements, if possible.

We can separate reputation cycles from epoch duration. However one reputation cycle is related to specific epoch. We can start once reputation cycle per epoch and isolate this cycle from new epoch events. Then there will be no data races and each reputation cycle will be finished based on the network configuration from this epoch.
